### ISSUE TITLE
vmstorage not mergeing into big part

### ISSUE THREAD

**Issue by Guoyuechuan** [VictoriaMetrics/VictoriaMetrics#1043]
Hi, I have three vmstorage nodes. All of them only store data in small parts and never merge small parts into big parts. I do not realize this problem until there is no free disk space. My data-writer writes the data for only four or five days with 3Mi ingestion rate. This disk usage does not make sense. Is there idea about this?

**Comment by hagen1778** [comment]
Hi @Guoyuechuan! VM requires free resource to complete merges, such as free disk space. There is a metric which detects if there are issues which prevent VM to perform merges - see more details here https://github.com/VictoriaMetrics/VictoriaMetrics/issues/686
I'd also recommend to use VM dashboards in order to track resource usage, merges and other useful metrics https://github.com/VictoriaMetrics/VictoriaMetrics#monitoring

**Comment by Guoyuechuan** [comment]
Thanks for the reply. I am using the Grafana Dashboard and have added the `vm_merge_need_free_disk_space` monitoring. The name of this metric is changed from the quoted issue.
Indeed, my disk space is not enough for merge parts. But is it possible to merge ealier before there is no enough disk space? Or the the imported data feature does not meet the criteria of merging into big parts? The data size is small and most of disk space is used by index file. It seems unreasonable that there is no a single big part during four days. Is there any idea about this?

**Comment by hagen1778** [comment]
The merging process happens in background and requires a free disk space and free CPU resources to perform it. Under some conditions, like frequent data deletion or lack of resources, merges could be delayed significantly. Have you tried to perform a forced merge? https://github.com/VictoriaMetrics/VictoriaMetrics/blob/bd11fd8f1dc12ebc5611b717360ed1cf93ec01aa/README.md#forced-merge

**Comment by Guoyuechuan** [comment]
I did not delete any data. Forced merging also failed. Here is the log.

![image](https://user-images.githubusercontent.com/20922937/106692648-d17b6680-660f-11eb-95e6-69da3ed87231.png)

For a single vmstorage node, the`vm_free_disk_space_bytes ` is 200G right now and the `vm_data_size_bytes ` is about 150G. The disk space should not be the problem. The node has 70 free CPUs. This is strange.

**Comment by Guoyuechuan** [comment]
Forced merging works.
I do not delete any data and the resource is sufficient for merging into big part. But the merging delayed. Is there any way to merge into big part automatically rather than manual forced merging under this situation?

**Comment by valyala** [comment]
@Guoyuechuan , which VictoriaMetrics version do you use? There were some fixes in the commit 9dcb18e03d1ef04901feca2232b609ca2803adb7 related to background merge process in v1.52.0 . Probably, these fixes prevent from starting big merges. Try downgrading to v1.51.0 and verifying whether big merges work as expected there. If you run vmstorage with version below v1.52.0, then try upgrading to the [latest available release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) and verify whether this fixes the issue.

As @hagen1778  already mentioned, the merge process cannot be started if there is no enough free disk space for storing the final merged part. The required disk space is estimated as the sum of sizes for the source parts that need to be merged. Probably you have no enough free space for big merges. It is recommended to keep at least 20% (30% is better) of free disk space at `-storageDataPath` directory. Disk space shortage may lead to increased number of data files and to vmstorage slowdown.

> most of disk space is used by index file

This looks suspicious. The size of `<-storageDataPath>/indexdb` directory (which contains index data for fast lookups of time series by label filters) should be much smaller than the size of `<-storageDataPath>/data` directory (which contains time series data) under normal conditions. If this isn't the case, then it is likely the stored data leads to too high churn rate or to too high number of unique time series. A time series is uniquely identified by its name plus a set of all its labels. If at least a single label value changes, then new time series is created. Every new time series increases index size, since indexes for all its labels are stored there. Take a look at `Active time series` and `Churn rate` graphs on [the official Grafana dashboard for VictoriaMetrics](https://victoriametrics.github.io/Cluster-VictoriaMetrics.html#monitoring) in order to determine these pathological cases. The source of high churn rate can be identified with the help of `/select/<accountID>/prometheus/api/v1/status/tsdb` page, which contains information about metric names and labels with the biggest number of distinct time series. See [these docs](https://prometheus.io/docs/prometheus/latest/querying/api/#tsdb-stats) for more information.

**Comment by Guoyuechuan** [comment]
Thanks for the detailed explanation.
I am using version `vmstorage-20210113-113338-tags-v1.52.0-cluster-0-gd23d6d4bf`. My free disk space is larger than the size of <-storageDataPath> and CPU is also sufficient. I will try to downgrade the version and see if it changes.
The churn rate is like
![image](https://user-images.githubusercontent.com/20922937/106833226-24b3ee80-66ce-11eb-9d3d-b6edf3594286.png)
The active series are
![image](https://user-images.githubusercontent.com/20922937/106833294-3ac1af00-66ce-11eb-9af1-30eb7f085389.png)
These seem like normal.

**Comment by hagen1778** [comment]
Hi @Guoyuechuan! Do you still experience issues with part merges?
